ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs

```diff
@@ -203,15 +206,53 @@ getBlockWithPoint db BlockPoint { withHash = hash, atSlot = slot } =
     --   non-existent after)
     -- - If the block does not exist in the database at all, we'd have done two
     --   reads before returning 'Nothing', but both of those reads are cheap
-    getBlockWithHash (Right slot) >>= \case
+    --
+    -- Note that there is one exceptional scenario: the point refers to the
+    -- EBB at the ImmutableDB's tip. This means that the regular block with
+    -- the same slot number as the EBB hasn't been added to the ImmutableDB
+    -- yet. If we first try to read the regular block at that slot, we would
+    -- get a 'ReadFutureSlotError' because we're trying to read a block in the
+    -- future. If we first read the EBB at that slot (the corresponding epoch,
+    -- in fact), then we won't get the 'ReadFutureSlotError'.
+    tipIsEBB >>= \case
+      Just tipSlot
+        | tipSlot == slot
+        -> getEBBThenBlock
+      _ -> getBlockThenEBB
+  where
+    EpochInfo{..} = epochInfo db
+
+    -- | If there's an EBB at the tip of the ImmutableDB, return its 'SlotNo'.
+    tipIsEBB :: m (Maybe SlotNo)
+    tipIsEBB = withDB db $ \imm -> ImmDB.getTip imm >>= \case
+      Tip (Left epochNo) -> Just <$> epochInfoFirst epochNo
+      Tip (Right _)      -> return Nothing
+      TipGen             -> return Nothing
+
+    -- | First try to read the block at the slot, if the block's hash doesn't
```

Comment:
No need to fall back to getting a block: we _know_ we are getting an EBB at this point. We should just check the hash, if it matches, return `Just` the EBB; if it doens't match, return `Nothing`; if we _fail_ to get the EBB, something went horribly wrong and we can throw an `ImmDbMissingBlock` exception.